languages/rust/oso/src/host/to_polar.rs

```diff
@@ -31,40 +31,41 @@ use crate::PolarValue;
 /// `ToPolar` implementors must also be concrete, sized types without
 /// any borrows.
 pub trait ToPolar: Send + Sync + Sized + 'static {
-    fn to_polar_value(self, host: &mut Host) -> Value {
-        let instance = Instance::new(self);
-        let instance = host.cache_instance(instance, None);
-        Value::ExternalInstance(ExternalInstance {
-            constructor: None,
-            repr: Some(std::any::type_name::<Self>().to_owned()),
-            instance_id: instance,
-        })
+    fn to_polar(self) -> PolarValue {
+        PolarValue::new_from_instance(self)
     }
+}
+
+impl<C: crate::PolarClass + Send + Sync> ToPolar for C {
+    fn to_polar(self) -> PolarValue {
+        DEFAULT_CLASSES
```

Comment:
This is the worst part of the change. We used to be able to just in time register the class on the host. But now we have this `T -> PolarValue -> Term` two step process, we lose the type in step 1, but only have the host in step 2.

This keeps the old behaviour (classes work without registering them), but introduces this global lazy static to make that work.

Reply:
What was wrong with passing host into each of these conversion functions? I haven't spend that much time reasoning through this, but would a global behave correctly if there are multiple instances of oso? We have tried to remove uses of globals in most of the other language libraries.

Reply:
I think it was fine, as long as this was a fully internal trait. But I think the really cool thing we get by having an intermediate `PolarValue` type, is that it's actually reasonable for somebody to implement `ToPolar` who doesn't need to understand any of the internal details. E.g. you might have a newtype `NonZero(u32)` that you want to implement to/from polar for. Not needing to know what the hell `Host` is seems like a big win.